beaconChain/lib/stateTransition/processBlock.ts

```diff
@@ -1,9 +1,391 @@
+  const randaoRevealVerified = blsVerify(
+    proposer.pubkey,
+    intToBytes(currentEpoch, 32),
+    block.randaoReveal,
+    getDomain(state.fork, currentEpoch, Domain.PROPOSAL),
+  );
+  if (!randaoRevealVerified) {
+    throw new Error();
+  }
+  state.latestRandaoMixes[currentEpoch.modn(LATEST_RANDAO_MIXES_LENGTH)] =
+    xor(getRandaoMix(state, currentEpoch), hash(block.randaoReveal));
+
+  // Eth1 data
+  let eth1DataVote: Eth1DataVote = state.eth1DataVotes.find((vote) =>
+    vote.eth1Data.blockHash.equals(block.eth1Data.blockHash) &&
+    vote.eth1Data.depositRoot.equals(block.eth1Data.depositRoot));
+
+  if (eth1DataVote) {
+    eth1DataVote.voteCount = eth1DataVote.voteCount.addn(1);
+  } else {
+    state.eth1DataVotes.push({
+      eth1Data: block.eth1Data,
+      voteCount: new BN(1),
+    });
+  }
+
+  // Transactions
+
+  // Proposer slashings
+  if (!(block.body.proposerSlashings.length <= MAX_PROPOSER_SLASHINGS)) {
+    throw new Error();
+  }
+  for (const proposerSlashing of block.body.proposerSlashings) {
+    const proposer: Validator =
+      state.validatorRegistry[proposerSlashing.proposerIndex.toNumber()];
+
+    if (!proposerSlashing.proposalData1.slot.eq(proposerSlashing.proposalData2.slot)) {
+      throw new Error();
+    }
+    if (!proposerSlashing.proposalData1.shard.eq(proposerSlashing.proposalData2.shard)) {
+      throw new Error();
+    }
+    if (!proposerSlashing.proposalData1.blockRoot.equals(proposerSlashing.proposalData2.blockRoot)) {
+      throw new Error();
+    }
+    if (!(proposer.slashedEpoch.gt(currentEpoch))) {
+      throw new Error();
+    }
+    const proposalData1Verified = blsVerify(
+      proposer.pubkey,
+      treeHash(proposerSlashing.proposalData1),
+      proposerSlashing.proposalSignature1,
+      getDomain(state.fork, slotToEpoch(proposerSlashing.proposalData1.slot), Domain.PROPOSAL)
+    );
+    if (!proposalData1Verified) {
+      throw new Error();
+    }
+    const proposalData2Verified = blsVerify(
+      proposer.pubkey,
+      treeHash(proposerSlashing.proposalData2),
+      proposerSlashing.proposalSignature2,
+      getDomain(state.fork, slotToEpoch(proposerSlashing.proposalData2.slot), Domain.PROPOSAL)
+    );
+    if (!proposalData2Verified) {
+      throw new Error();
+    }
+    slashValidator(state, proposerSlashing.proposerIndex);
+  }
+
+  // Attester slashings
+  if (!(block.body.attesterSlashings.length <= MAX_ATTESTER_SLASHINGS)) {
+    throw new Error();
+  }
+  for (const attesterSlashing of block.body.attesterSlashings) {
+    const slashableAttestation1 = attesterSlashing.slashableAttestation1;
+    const slashableAttestation2 = attesterSlashing.slashableAttestation2;
+
+    if (serialize(slashableAttestation1.data, AttestationData).eq(serialize(slashableAttestation2.data, AttestationData))) {
+      throw new Error();
+    }
+    if (isDoubleVote(slashableAttestation1.data, slashableAttestation2.data) ||
```

Comment:
Please confirm that the spec is wrong here. I read it as one of these must be true: https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#attester-slashings-1

Reply:
You're totally right. I'd like to move to an "assert-style" for all of these checks so it should be easier to read (and less verbose).